In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 

In [2]:
df = pd.read_csv('../data/processed/main_train.csv')

In [3]:
df.head(5)

Unnamed: 0  case_id date_decision_x  MONTH_x  WEEK_NUM_x  target_x  \
0           0        0      2019-01-03   201901           0         0   
1           1        1      2019-01-03   201901           0         0   
2           2        2      2019-01-04   201901           0         0   
3           3        3      2019-01-03   201901           0         0   
4           4        4      2019-01-04   201901           0         1   

   amtdebitincoming_4809443A  amtdebitoutgoing_4809440A  \
0                        NaN                        NaN   
1                        NaN                        NaN   
2                        NaN                        NaN   
3                        NaN                        NaN   
4                        NaN                        NaN   

   amtdepositbalance_4809441A  amtdepositincoming_4809444A  ...  maininc_215A  \
0                         NaN                          NaN  ...           NaN   
1                         NaN                          NaN  ...           NaN   
2                         NaN                          NaN  ...           NaN   
3                         NaN                          NaN  ...           NaN   
4                         NaN                          NaN  ...           NaN   

   downpmt_116A  avgdbddpdlast24m_3658932P  empl_employedtotal_800L  \
0           0.0                        NaN                MORE_FIVE   
1           0.0                        NaN                MORE_FIVE   
2           0.0                        NaN                MORE_FIVE   
3           0.0                        NaN                MORE_FIVE   
4           0.0                        NaN                MORE_FIVE   

  empl_industry_691L  date_decision   MONTH  WEEK_NUM  target  \
0              OTHER     2019-01-03  201901         0       0   
1              OTHER     2019-01-03  201901         0       0   
2              OTHER     2019-01-04  201901         0       0   
3              OTHER     2019-01-03  201901         0       0   
4              OTHER     2019-01-04  201901         0       1   

   avgdbddpdlast3m_4187120P  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN  

[5 rows x 31 columns]

In [4]:
df.drop(columns= ['Unnamed: 0'], inplace = True)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1526659 entries, 0 to 1526658
Data columns (total 30 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   case_id                      1526659 non-null  int64  
 1   date_decision_x              1526659 non-null  object 
 2   MONTH_x                      1526659 non-null  int64  
 3   WEEK_NUM_x                   1526659 non-null  int64  
 4   target_x                     1526659 non-null  int64  
 5   amtdebitincoming_4809443A    51109 non-null    float64
 6   amtdebitoutgoing_4809440A    51109 non-null    float64
 7   amtdepositbalance_4809441A   51109 non-null    float64
 8   amtdepositincoming_4809444A  51109 non-null    float64
 9   amtdepositoutgoing_4809442A  51109 non-null    float64
 10  annuity_780A                 1526659 non-null  float64
 11  cntincpaycont9m_3716944L     1071469 non-null  float64
 12  amount_416A                  105111 non-nu

In [6]:
df.isna().sum()

case_id                              0
date_decision_x                      0
MONTH_x                              0
WEEK_NUM_x                           0
target_x                             0
amtdebitincoming_4809443A      1475550
amtdebitoutgoing_4809440A      1475550
amtdepositbalance_4809441A     1475550
amtdepositincoming_4809444A    1475550
amtdepositoutgoing_4809442A    1475550
annuity_780A                         0
cntincpaycont9m_3716944L        455190
amount_416A                    1421548
date_decision_y                      0
MONTH_y                              0
WEEK_NUM_y                           0
target_y                             0
cntpmts24_3658933L              460822
commnoinclast6m_3546845L        343375
equalityempfrom_62L            1488847
maininc_215A                    511255
downpmt_116A                         0
avgdbddpdlast24m_3658932P       613202
empl_employedtotal_800L        1004548
empl_industry_691L             1004548
date_decision            

In [9]:
df.shape

(1526659, 30)

Based on the info we have on our dataset currently. We decided that some of the columns have way too much null values. We did some quick calculations below to determine exactly how much null values are present in the full dataset and from there made an executive decision to drop them because we think they would not provide us with any meaningful information. 

Here are the columns we will be dropping below as well as its calculations.

amtdebitincoming_4809443A    51109  / 1526659 = 3.3% non null value out of the whole column (Not sufficient for analysis)
amtdebitoutgoing_4809440A    51109  / 1526659 = 3.3% non null value out of the whole column (Not sufficient for analysis)
amtdepositbalance_4809441A   51109  / 1526659 = 3.3% non null value out of the whole column (Not sufficient for analysis)
amtdepositincoming_4809444A  51109  / 1526659 = 3.3% non null value out of the whole column (Not sufficient for analysis)
amtdepositoutgoing_4809442A  51109  / 1526659 = 3.3% non null value out of the whole column (Not sufficient for analysis)
amount_416A                  105111 / 1526659 = 
equalityempfrom_62L          37812  / 1526659 = 



We want to oversample this column based on the exact ratio of the values that currently exist in the column

empl_employedtotal_800L


In [7]:
df.describe()

case_id       MONTH_x    WEEK_NUM_x      target_x  \
count  1.526659e+06  1.526659e+06  1.526659e+06  1.526659e+06   
mean   1.286077e+06  2.019363e+05  4.076904e+01  3.143728e-02   
std    7.189466e+05  4.473597e+01  2.379798e+01  1.744964e-01   
min    0.000000e+00  2.019010e+05  0.000000e+00  0.000000e+00   
25%    7.661975e+05  2.019060e+05  2.300000e+01  0.000000e+00   
50%    1.357358e+06  2.019100e+05  4.000000e+01  0.000000e+00   
75%    1.739022e+06  2.020010e+05  5.500000e+01  0.000000e+00   
max    2.703454e+06  2.020100e+05  9.100000e+01  1.000000e+00   

       amtdebitincoming_4809443A  amtdebitoutgoing_4809440A  \
count               5.110900e+04               5.110900e+04   
mean                7.552902e+03               7.462384e+03   
std                 3.462571e+04               3.506529e+04   
min                 0.000000e+00               0.000000e+00   
25%                 0.000000e+00               0.000000e+00   
50%                 0.000000e+00               0.000000e+00   
75%                 8.000000e+03               7.740000e+03   
max                 4.957852e+06               5.168004e+06   

       amtdepositbalance_4809441A  amtdepositincoming_4809444A  \
count                5.110900e+04                 5.110900e+04   
mean                 9.967413e+03                 2.949396e+03   
std                  8.939342e+04                 4.146773e+04   
min                 -3.357180e+05                 0.000000e+00   
25%                  0.000000e+00                 0.000000e+00   
50%                  0.000000e+00                 0.000000e+00   
75%                  2.880000e+02                 0.000000e+00   
max                  4.256314e+06                 4.180150e+06   

       amtdepositoutgoing_4809442A  annuity_780A  ...      target_y  \
count                 5.110900e+04  1.526659e+06  ...  1.526659e+06   
mean                  3.586875e+03  4.039207e+03  ...  3.143728e-02   
std                   4.827494e+04  3.006608e+03  ...  1.744964e-01   
min                   0.000000e+00  8.080000e+01  ...  0.000000e+00   
25%                   0.000000e+00  1.967600e+03  ...  0.000000e+00   
50%                   1.800000e+00  3.151800e+03  ...  0.000000e+00   
75%                   5.400000e+00  5.231400e+03  ...  0.000000e+00   
max                   4.622918e+06  1.060070e+05  ...  1.000000e+00   

       cntpmts24_3658933L  commnoinclast6m_3546845L  maininc_215A  \
count        1.065837e+06                 1183284.0  1.015404e+06   
mean         1.063064e+01                       0.0  4.948565e+04   
std          7.888573e+00                       0.0  3.226414e+04   
min          0.000000e+00                       0.0  0.000000e+00   
25%          4.000000e+00                       0.0  3.000000e+04   
50%          1.000000e+01                       0.0  4.000000e+04   
75%          1.700000e+01                       0.0  6.000000e+04   
max          2.500000e+01                       0.0  2.000000e+05   

       downpmt_116A  avgdbddpdlast24m_3658932P         MONTH      WEEK_NUM  \
count  1.526659e+06              913457.000000  1.526659e+06  1.526659e+06   
mean   5.523810e+02                  30.098977  2.019363e+05  4.076904e+01   
std    4.100824e+03                 317.684627  4.473597e+01  2.379798e+01   
min    0.000000e+00               -1220.000000  2.019010e+05  0.000000e+00   
25%    0.000000e+00                 -11.000000  2.019060e+05  2.300000e+01   
50%    0.000000e+00                  -5.000000  2.019100e+05  4.000000e+01   
75%    0.000000e+00                  -1.000000  2.020010e+05  5.500000e+01   
max    4.618674e+05                4750.000000  2.020100e+05  9.100000e+01   

             target  avgdbddpdlast3m_4187120P  
count  1.526659e+06             578415.000000  
mean   3.143728e-02                 29.494342  
std    1.744964e-01                367.273913  
min    0.000000e+00               -908.000000  
25%    0.000000e+00                -11.000000  
50%    0.0

In [ ]:
#Our data is way too large and computationally heavy to conduct EDA. To speed up the process I decided to use random sampling on my data.
#Since our goal is to initially explore data in order to identify trends and patterns a smaller sample size could work.

sampled_df = df.sample(n=50000, random_state=42)

## Univariate Analysis

Let's start with a simple univariate analysis to identify outliers and distribution that is present in our dataset.

In [18]:
# Calculating the skewness of each feature
to_drop = ['case_id', "date_decision_x", "date_decision_y", "equalityempfrom_62L","empl_employedtotal_800L","empl_industry_691L","date_decision", "target"]
other_columns = df.drop(to_drop, axis=1).columns
feature_skewness = df[other_columns].skew()

# Plotting histograms for each feature
plt.figure(figsize=(15, 10))
for i, column in enumerate(other_columns, 1):  # Excluding the target variable
    plt.subplot(5, 5, i)
    sns.histplot(df[column], kde=True)
    plt.title(f'{column} (Skewness: {feature_skewness[column]:.2f})')

plt.tight_layout()
plt.show()

feature_skewness